base-console-app-cert/Program.cs

```diff
@@ -0,0 +1,131 @@
+        private static HttpClient _httpClient;
+
+        static void Main(string[] args)
+        {
+            // Load appsettings.json
+            var config = LoadAppSettings();
+            if (null == config)
+            {
+                Console.WriteLine("Missing or invalid appsettings.json file. Please see README.md for configuration instructions.");
+                return;
+            }
+
+            //Query using Graph SDK (preferred when possible)
+            GraphServiceClient graphClient = GetAuthenticatedGraphClient(config);
+            List<QueryOption> options = new List<QueryOption>
+            {
+                new QueryOption("$top", "1")
+            };
+
+            var graphResult = graphClient.Users.Request(options).GetAsync().Result;
+            Console.WriteLine("Graph SDK Result");
+            Console.WriteLine(graphResult[0].DisplayName);
+
+            //Direct query using HTTPClient (for beta endpoint calls or not available in Graph SDK)
+            HttpClient httpClient = GetAuthenticatedHTTPClient(config);
+            Uri Uri = new Uri("https://graph.microsoft.com/v1.0/users?$top=1");
+            var httpResult = httpClient.GetStringAsync(Uri).Result;
+
+            Console.WriteLine("HTTP Result");
+            Console.WriteLine(httpResult);
+        }
+
+        private static GraphServiceClient GetAuthenticatedGraphClient(IConfigurationRoot config)
+        {
+            var authenticationProvider = CreateAuthorizationProvider(config);
+            _graphServiceClient = new GraphServiceClient(authenticationProvider);
+            return _graphServiceClient;
+        }
+
+        private static HttpClient GetAuthenticatedHTTPClient(IConfigurationRoot config)
+        {
+            var authenticationProvider = CreateAuthorizationProvider(config);
+            _httpClient = new HttpClient(new AuthHandler(authenticationProvider, new HttpClientHandler()));
+            return _httpClient;
+        }
+
+        private static IAuthenticationProvider CreateAuthorizationProvider(IConfigurationRoot config)
+        {
+            var clientId = config["applicationId"];
+            var certificateThumbprint = config["certificateThumbprint"];
+            var redirectUri = config["redirectUri"];
+            var authority = $"https://login.microsoftonline.com/{config["tenantId"]}/v2.0";
+
+            // defaulting to CurrentUser certificate store under My (Personal), change these if stored elsewhere
+            X509Certificate2 cert = GetCertificate(certificateThumbprint, StoreName.My, StoreLocation.CurrentUser);
+
+            //this specific scope means that application will default to what is defined in the application registration rather than using dynamic scopes
+            List<string> scopes = new List<string>();
+            scopes.Add("https://graph.microsoft.com/.default");
+
+            var cca = ConfidentialClientApplicationBuilder.Create(clientId)
```

Comment:
Why not have used `CreateWithOptions()` ? (with a slightly different format of the appsettings.json)

Reply:
@jmprieur Great suggestion.  The least amount of changes were implemented to update to the MSAL .Net v3 / v4 APIs.  Refactoring the original appsettings.json will take additional effort across each of the impacted projects.  Please open an issue (or I can if you'd like) to implement those changes.

Reply:
https://github.com/microsoftgraph/dotnetcore-console-sample/issues/51